src/OgAccess.php

```diff
@@ -0,0 +1,278 @@
+  }
+
+  /**
+   * Check if a user has access to a permission on a certain entity context.
+   *
+   * @param string $operation
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity object.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   (optional) The user object. If empty the current user will be used.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   Returns TRUE if the user has access to the permission, otherwise FALSE, or
+   *   if the entity is not in OG context, function _will return NULL. This allows
+   *   a distinction between FALSE - no access, and NULL - no access as no OG
+   *   context found.
+   */
+  public static function userAccessEntity($operation, EntityInterface $entity, AccountInterface $user = NULL) {
+    $result = AccessResult::neutral()
+      ->addCacheableDependency($entity)
+      ->addCacheableDependency($user);
+
+    // Entity isn't saved yet.
+    if ($entity->isNew()) {
+      return $result;
+    }
+
+    $entity_type = $entity->getEntityTypeId();
+    $bundle = $entity->bundle();
+
+    $is_group_content = Og::isGroupContent($entity_type, $bundle);
+
+    if (Og::isGroup($entity_type, $bundle)) {
+      $user_access = static::userAccess($entity, $operation, $user);
+      if ($user_access->isAllowed()) {
+        return $user_access;
+      }
+      else {
+        // An entity can be a group and group content in the same time. The
+        // group didn't return TRUE, but the user still might have access to the
+        // permission in group content context. So instead of retuning a deny
+        // here, we set the result, that might change if an access is found.
+        // @TODO Change something about caching and user?
+        $result = AccessResult::forbidden()->addCacheableDependency($entity);
+      }
+    }
+
+    if ($is_group_content && $result = Og::getEntityGroups($entity)) {
+      foreach ($result as $groups) {
+        foreach ($groups as $group) {
+          $user_access = static::userAccess($group, $operation, $user);
+          if ($user_access->isAllowed()) {
+            return $user_access;
+          }
+        }
+      }
+      // @TODO Change something about caching and user?
+      return AccessResult::forbidden()->addCacheableDependency($entity);
+    }
+
+    // Either the user didn't have permission, or the entity might be an
+    // orphaned group content.
+    return $result;
+  }
+
+  /**
+   * Set the permissions in the static cache.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $group
+   *   The entity object.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user object.
+   * @param bool $pre_alter $type
+   *   Determines if the type of permissions is pre-alter or post-alter.
+   * @param array $permissions
+   *   Array of permissions to set.
+   */
+  public static function setPermissionCache(EntityInterface $group, AccountInterface $user, $pre_alter, array $permissions) {
```

Comment:
question is btw what to do here -- this class is not initalized, we just use the static methods. I wonder what would be the best approach for that.